### ISSUE TITLE
usize and u64 mismatching

### ISSUE THREAD

**Issue by NullP01nt** [phil-opp/blog_os#302]
Following along with the development of the blog_os, rustc is not very happy with me. I ended up in the post about remapping the kernel and set to work on some of the memory subsystem code.

error: non-scalar cast: `x86_64::PhysicalAddress` as `usize`  --> src/memory/paging/mod.rs:91:52
`let backup = Frame::containing_address(unsafe { control_regs::cr3() } as usize);`

error[E0308]: mismatched types --> src/memory/paging/mod.rs:185:36
`assert!(section.addr % PAGE_SIZE == 0, "sections need to be page aligned");`
PAGE_SIZE was expected to be u64, but found to be usize.

error[E0277]: the trait bound `u64: core::ops::Rem<usize>` is not satisfied --> src/memory/paging/mod.rs:185:21
`assert!(section.addr % PAGE_SIZE == 0, "sections need to be page aligned");`
the trait 'core::ops::Rem<usize>' is not implemented for 'u64'.

I'm not quite sure how to fix this, I could make all usize type definitions simply be u64 type definitions since the target architecture is x86_64 anyway, but I'm not sure if that's the cleanest solution.

I'm using `rustc 1.17.0-nightly (cab4bff3d 2017-03-21)`

**Comment by zarevucky** [comment]
PhysicalAddress is defined as
`pub struct PhysicalAddress(pub u64);`
so you have to unwrap it first, then cast. Something like
`let PhysicalAddress(x) = unsafe { control_regs::cr3() }`
`let backup = Frame::containing_address(x as usize);`

Or you can make Frame::containing_address() accept x86_64::PhysicalAddress directly, and do the unwrapping inside.

**Comment by NullP01nt** [comment]
Thanks for clarifying zarevucky, I did a little bit of digging and came up with the following:
`let backup = Frame::containing_address(unsafe { control_regs::cr3().0 as usize } );`
Not sure if it's the safest way, but it extracts the u64 which is wrapped in the x86_64::PhysicalAddress struct. If anyone else has a better idea, let me know.

**Comment by zarevucky** [comment]
Yeah, that's basically the same thing that my version does, just using a different syntax.
